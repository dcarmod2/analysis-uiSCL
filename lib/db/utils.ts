import mapValues from 'lodash/mapValues'
import {ObjectID} from 'mongodb'

export function serializeCollection(obj) {
  if (obj instanceof Date) return obj.toISOString()
  if (obj instanceof ObjectID) return obj.toString()
  if (Array.isArray(obj)) return obj.map(serializeCollection)
  if (typeof obj === 'object') return mapValues(obj, serializeCollection)
  return obj
}
