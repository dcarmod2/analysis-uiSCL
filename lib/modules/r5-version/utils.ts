import {MINIMUM_R5_VERSION, VERSION_PARSE_REGEX} from './constants'

/**
 * Convert an r5 version to a number for comparing
 */
export function versionToNumber(version: string): number {
  const versionMatch = (version || '').match(VERSION_PARSE_REGEX)
  if (versionMatch) {
    const [, major, minor, patch, commit] = versionMatch

    // NB Assumes no major, minor, patch, or commit version exceeds 100
    return (
      parseInt(major) * 10000 +
      parseInt(minor) * 100 +
      parseInt(patch) +
      parseInt(commit || '0') / 100
    )
  }

  return 0
}

/**
 * Check if the r5 version is older than a given minimum.
 */
export function workerVersionInRange(
  version: string,
  min: string = MINIMUM_R5_VERSION,
  max = 'v1000.0.0'
): boolean {
  const versionNumber = versionToNumber(version)
  return (
    versionNumber >= versionToNumber(min) &&
    versionNumber <= versionToNumber(max)
  )
}

/**
 * Check if the r5 version is a patch version or in a given range
 */
export function workerVersionTestOrInRange(
  version: string,
  min: string,
  max?: string
): boolean {
  const isTestVersion = versionToNumber(version) % 1 !== 0
  return isTestVersion || workerVersionInRange(version, min, max)
}
