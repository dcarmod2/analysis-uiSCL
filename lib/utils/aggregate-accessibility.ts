import get from 'lodash/get'
import range from 'lodash/range'

const N_HISTOGRAM_BINS = 15

/**
 * Compute the accessibility, exported for use in comparison weighted average
 * accessibility.
 * @param weights a grid containing the weights (e.g. population)
 * @param aggregationArea a grid containing the aggregation area, with the
 *        percentage the aggregation area overlaps each cell, scaled between 0
 *        (no overlap) and 100,000 (full coverage)
 * @param accessibility a grid containing the accessibility (i.e. the regional
 *        analysis results)
 */
export default function getAggregateAccessibility(
  accessibility,
  aggregationArea,
  weights
) {
  const aa = get(aggregationArea, 'grid')
  if (weights == null || aa == null || accessibility == null) {
    return
  }

  if (accessibility.zoom !== weights.zoom || accessibility.zoom !== aa.zoom) {
    console.error('Zoom levels of all grids in weighted average must match!')
  }

  // find max and min value for histogram
  let maxAccessibility = 0
  let minAccessibility = Infinity

  // Pass one: find min and max from aggregationArea
  // aggregationArea is probably the smallest grid and is guaranteed to contain
  // all pixels
  for (let y = 0, pixel = 0; y < aa.height; y++) {
    // Transform to coordinates in other grid; all grids have same zoom, so
    // translate based on the difference in where the edges of the grid are
    const accessY = y + aa.north - accessibility.north
    for (let x = 0; x < aa.width; x++, pixel++) {
      if (aa.data[pixel] > 0) {
        const accessX = x + aa.west - accessibility.west
        const accessibilityThisCell = accessibility.contains(accessX, accessY)
          ? accessibility.data[accessY * accessibility.width + accessX]
          : 0

        minAccessibility = Math.min(accessibilityThisCell, minAccessibility)
        maxAccessibility = Math.max(accessibilityThisCell, maxAccessibility)
      }
    }
  }

  const spread = maxAccessibility - minAccessibility
  const bins = range(N_HISTOGRAM_BINS).map(() => 0)

  let sumOfWeights = 0
  // weight and accessibility value for each cell, used to compute percentiles
  const weightedAccessibility = []

  // pass two: accumulate into bins
  for (let y = 0, pixel = 0; y < aa.height; y++) {
    const accessY = y + aa.north - accessibility.north
    const weightY = y + aa.north - weights.north
    for (let x = 0; x < aa.width; x++, pixel++) {
      if (aa.data[pixel] > 0) {
        const accessX = x + aa.west - accessibility.west
        const accessibilityThisCell = accessibility.contains(accessX, accessY)
          ? accessibility.data[accessY * accessibility.width + accessX]
          : 0

        let binIndex = Math.floor(
          ((accessibilityThisCell - minAccessibility) / spread) *
            N_HISTOGRAM_BINS
        )

        if (binIndex > N_HISTOGRAM_BINS) console.error('invalid bin!')

        // avoid putting the highest accessibility cell beyond the histogram
        if (binIndex === N_HISTOGRAM_BINS) binIndex--

        // how much of this cell is within the aggregationArea
        const aggregationAreaWeightThisCell = aa.data[pixel] / 100000

        const weightX = x + aa.west - weights.west
        const weightThisCell = weights.contains(weightX, weightY)
          ? weights.data[weightY * weights.width + weightX]
          : 0

        const weight = aggregationAreaWeightThisCell * weightThisCell
        bins[binIndex] += aggregationAreaWeightThisCell * weightThisCell

        sumOfWeights += weight
        weightedAccessibility.push({
          weight,
          accessibility: accessibilityThisCell
        })
      }
    }
  }

  // Figure out percentiles
  let cumulativeWeight = 0
  // Preinitialize with null so 50th percentile is percentiles[50].
  // There is no zeroth percentile
  const percentiles = [0]
  const weightPerPercentile = sumOfWeights / 100
  let expectedWeight = weightPerPercentile
  weightedAccessibility.sort((a, b) => a.accessibility - b.accessibility)

  // we calculate and return an old-style weighted average for comparison purposes.
  let sumOfWeightedAccessibility = 0

  // We loop over the sorted weighted accessibility values. When we cross a
  // percentile boundary, we push that percentile onto the list
  weightedAccessibility.forEach(({accessibility, weight}) => {
    sumOfWeightedAccessibility += accessibility * weight

    cumulativeWeight += weight

    // while not if, possible to cross many percentiles with one high-density cell
    while (cumulativeWeight > expectedWeight) {
      // There is a percentile value somewhere within the values at this sample
      // We use > above, not >=, so that a percentile value that falls exactly
      // between two accessibility values will get the higher value, e.g. if 10
      // people have accessibility 50 and ten people have accessibility 60, we
      // want 60. This is because in the UI we say "x% of <people> have
      // accessibility y or higher," so we want y to be as high as possible
      // while still being a true statement.
      percentiles.push(accessibility)
      expectedWeight += weightPerPercentile
    }
  })

  // includes values 1 - 99
  // When the resolution of the data is low the last item may only define (say)
  // the 90th percentile, all remaining percentiles are just the max. Also
  // handles the case when there is no accessibility within the aggregation
  // area by returning all zeros.
  while (percentiles.length < 99) {
    percentiles.push(
      weightedAccessibility.length > 0
        ? weightedAccessibility[weightedAccessibility.length - 1].accessibility
        : 0
    )
  }

  const binWidth = spread / N_HISTOGRAM_BINS

  return {
    minAccessibility,
    maxAccessibility,
    percentiles,
    weightedAverage: sumOfWeightedAccessibility / sumOfWeights,
    bins: bins.map((value, idx) => ({
      value,
      min: minAccessibility + idx * binWidth,
      max: minAccessibility + (idx + 1) * binWidth
    }))
  }
}
