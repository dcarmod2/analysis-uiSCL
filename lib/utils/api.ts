// Query values are typed as string OR string[]
export function getQueryAsString(p: string | string[]): string {
  return Array.isArray(p) ? p[0] : p
}

// Get a query as an object
export function getQueryAsObject(
  p: string | string[]
): Record<string, unknown> {
  try {
    return JSON.parse(getQueryAsString(p))
  } catch (e) {
    return {}
  }
}

// Convert an error to POJO for sending over the wire
export function errorToPOJO(e: Error): Record<string, unknown> {
  const obj = {}
  Object.getOwnPropertyNames(e).forEach((k) => {
    obj[k] = e[k]
  })
  return obj
}
