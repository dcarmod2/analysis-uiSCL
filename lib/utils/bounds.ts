import lonlat from '@conveyal/lonlat'

import L from 'lib/leaflet'
import message from 'lib/message'

import reprojectCoordinates from './reproject-coordinates'

const ANTIMERIDIAN_WARNING =
  'We do not currently support bounds that cross the antimeridian.'

type Bounds = {
  north: number
  south: number
  east: number
  west: number
  toString: () => string
}

export function toLatLngBounds(
  bounds: Bounds | L.LatLngBounds
): L.LatLngBounds {
  // Check if it's already a leaflet bounds object
  if (bounds instanceof L.LatLngBounds) return bounds
  return new L.LatLngBounds([
    [bounds.north, bounds.west],
    [bounds.south, bounds.east]
  ])
}

/**
 * NB: We do not currently support crossing the antimeridian.
 */
export function fromLatLngBounds(bounds: L.LatLngBounds) {
  // Don't change the original bounds object
  let b = new L.LatLngBounds(bounds.getSouthWest(), bounds.getNorthEast())

  if (b.getWest() < -180) {
    b = new L.LatLngBounds(new L.LatLng(b.getSouth(), -180), b.getNorthEast())
    window.alert(message('antimeridian', ANTIMERIDIAN_WARNING))
  }

  if (b.getEast() > 180) {
    b = new L.LatLngBounds(new L.LatLng(b.getSouth(), 180), b.getNorthEast())
    window.alert(message('antimeridian', ANTIMERIDIAN_WARNING))
  }

  const bboxString = b.toBBoxString()
  return {
    north: b.getNorth(),
    south: b.getSouth(),
    east: b.getEast(),
    west: b.getWest(),
    toString: () => bboxString
  }
}

/**
 * Reproject bounds to the grid size that we use.
 */
export function reprojectBounds(bounds) {
  return new L.LatLngBounds(
    lonlat.toLeaflet(reprojectCoordinates(bounds.getSouthWest())),
    lonlat.toLeaflet(reprojectCoordinates(bounds.getNorthEast()))
  )
}
