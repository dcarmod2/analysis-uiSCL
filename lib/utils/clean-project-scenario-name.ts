//
import message from 'lib/message'

import {UNDEFINED_PROJECT_NAME} from '../constants'

export default function cleanProjectScenarioName(project, scenarioIndex) {
  if (!project) {
    return UNDEFINED_PROJECT_NAME
  }
  const scenarioName =
    project.variants[scenarioIndex] || message('variant.baseline')
  const description = `${project.name}-${scenarioName}`
  return description.replace(/[^a-zA-Z0-9]/, '-')
}
