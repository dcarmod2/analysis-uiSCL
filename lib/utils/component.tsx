import {ThemeProvider} from '@chakra-ui/core'
import {expect} from '@jest/globals'
import enzyme from 'enzyme'
import {Router} from 'next/router'
import {RouterContext} from 'next/dist/next-server/lib/router-context'
import React from 'react'
import {Map} from 'react-leaflet'
import {Provider} from 'react-redux'

import {makeMockStore, mockStores} from './mock-data'

// Next.js uses Math.random to generate unique ids for pages that get captured in snapshots.
global.Math.random = () => 0.5

// Set __NEXT_DATA__ (required by Router)
window.__NEXT_DATA__ = {} as any
const router = new Router('page', {}, 'as-path', {} as any)

const getName = (c) => c.displayName || c.name

export function testComponent(
  TestComponent,
  props = {},
  initialData = mockStores.init
) {
  const store = makeMockStore(initialData)
  const Wrapping = (props) => (
    <Provider store={store}>
      <ThemeProvider>
        <RouterContext.Provider value={router}>
          {props.children}
        </RouterContext.Provider>
      </ThemeProvider>
    </Provider>
  )

  const mount = () =>
    enzyme.mount(<TestComponent {...props} />, {
      wrappingComponent: Wrapping
    })

  // TODO: Dive down through the `TestComponent` automatically
  const shallow = () =>
    enzyme.shallow(<TestComponent {...props} />, {
      wrappingComponent: Wrapping
    })

  // For debugging purposes (this is not the original component!)
  const name = `WithContext(${getName(TestComponent)})`

  return {mount, name, router, shallow, store}
}

/**
 * Rename to `wrapComponent`. TODO replace all `testComponent`
 */
export const wrapComponent = testComponent

/**
 * Helper function to create quick snapshot tests for components.
 */
export function testAndSnapshot(C, p, id?: any) {
  test(`Component ${getName(C)} snapshot(mount)`, () => {
    const c = testComponent(C, p, id)
    const t = c.mount()
    expect(t).toMatchSnapshot()
    t.unmount()
  })
}

/**
 * Test and snapshot map component
 */
export function wrapMapComponent(C, p, id?: any) {
  const MapWrapper = (props) => (
    <Map>
      <C {...props} />
    </Map>
  )
  MapWrapper.displayName = `WithMap(${getName(C)})`
  return testComponent(MapWrapper, p, id)
}

function snapshotTest(wc, type) {
  test(`Component ${getName(wc)} snapshot(${type})`, () => {
    const s = wc[`${type}`]()
    expect(s).toMatchSnapshot()
    s.unmount()
  })
}

/**
 * Shallow snapshot test.
 */
export function shallowSnapshot(wc) {
  snapshotTest(wc, 'shallow')
}

/**
 * Deep snapshot test.
 */
export function deepSnapshot(wc) {
  snapshotTest(wc, 'mount')
}
