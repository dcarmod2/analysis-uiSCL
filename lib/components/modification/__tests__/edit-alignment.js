import {testAndSnapshot} from 'lib/utils/component'
import {mockModification} from 'lib/utils/mock-data'
import EditAlignment from '../edit-alignment'

const props = {
  disabled: false,
  extendFromEnd: false,
  mapState: {
    modificationId: mockModification._id,
    state: null
  },
  modification: mockModification,
  allStops: [],
  numberOfStops: 16,
  segmentDistances: [45.22],
  update: jest.fn()
}

testAndSnapshot(EditAlignment, props)
