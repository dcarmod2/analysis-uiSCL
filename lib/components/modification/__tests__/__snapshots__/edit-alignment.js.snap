// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component EditAlignment snapshot(mount) 1`] = `
<EditAlignment
  allStops={Array []}
  disabled={false}
  extendFromEnd={false}
  mapState={
    Object {
      "modificationId": "1234",
      "state": null,
    }
  }
  modification={
    Object {
      "_id": "1234",
      "accessGroup": "accessGroupName",
      "bidirectional": false,
      "createdAt": "2017-11-08T12:30:50.133Z",
      "createdBy": "person@gmail.com",
      "description": "Mock description",
      "entries": Array [],
      "feed": "1",
      "name": "Test Modification",
      "nonce": "12345",
      "projectId": "1",
      "routes": Array [
        "route1",
      ],
      "segments": Array [
        Object {
          "fromStopId": "1",
          "geometry": Object {
            "coordinates": Array [
              Array [
                -122.0246,
                36.9707,
              ],
              Array [
                -122.0279,
                37.049,
              ],
              Array [
                -121.9799,
                37.2299,
              ],
              Array [
                -121.9445,
                37.324,
              ],
              Array [
                -121.936,
                37.353,
              ],
              Array [
                -121.924,
                37.365,
              ],
            ],
            "type": "LineString",
          },
          "spacing": 3000,
          "stopAtEnd": false,
          "stopAtStart": false,
          "toStopId": null,
        },
      ],
      "timetables": Array [
        Object {
          "_id": "timetable-id",
          "dwellTime": 10,
          "dwellTimes": Array [
            12,
          ],
          "endTime": 57600,
          "exactTimes": false,
          "friday": true,
          "headwaySecs": 900,
          "modificationId": "1234",
          "modificationName": "Test Modification",
          "monday": true,
          "name": "Test timetable",
          "patternTrips": Array [
            "abcd",
          ],
          "phaseAtStop": "",
          "phaseFromStop": "",
          "phaseFromTimetable": "",
          "phaseSeconds": 300,
          "saturday": false,
          "segmentSpeeds": Array [],
          "sourceTrip": "abcd",
          "startTime": 28800,
          "sunday": false,
          "thursday": true,
          "tuesday": true,
          "wednesday": true,
        },
      ],
      "trips": Array [
        "abcd",
      ],
      "type": "add-trip-pattern",
      "updatedAt": "2017-11-08T12:31:19.602Z",
      "updatedBy": "person@gmail.com",
      "variants": Array [
        true,
      ],
    }
  }
  numberOfStops={16}
  segmentDistances={
    Array [
      45.22,
    ]
  }
  update={[MockFunction]}
>
  <ForwardRef
    allStops={Array []}
    disabled={false}
    extendFromEnd={false}
    mapState={
      Object {
        "modificationId": "1234",
        "state": null,
      }
    }
    segmentDistances={
      Array [
        45.22,
      ]
    }
    spacing={4}
  >
    <Flex
      allStops={Array []}
      direction="column"
      disabled={false}
      extendFromEnd={false}
      mapState={
        Object {
          "modificationId": "1234",
          "state": null,
        }
      }
      segmentDistances={
        Array [
          45.22,
        ]
      }
    >
      <Box
        allStops={Array []}
        disabled={false}
        display="flex"
        extendFromEnd={false}
        flexDirection="column"
        mapState={
          Object {
            "modificationId": "1234",
            "state": null,
          }
        }
        segmentDistances={
          Array [
            45.22,
          ]
        }
      >
        <div
          className="css-j7qwjs"
          disabled={false}
        >
          <Heading
            key=".1"
            mb={4}
            size="sm"
          >
            <Box
              as="h2"
              fontFamily="heading"
              fontSize="md"
              fontWeight="bold"
              lineHeight="shorter"
              mb={4}
            >
              <h2
                className="css-35hltn"
              >
                Route Geometry
              </h2>
            </Box>
          </Heading>
          <Text
            key=".3"
            mb={4}
          >
            <Box
              as="p"
              fontFamily="body"
              mb={4}
            >
              <p
                className="css-zwncle"
              >
                16 stops over 45.21 km
              </p>
            </Box>
          </Text>
          <Button
            isFullWidth={true}
            key=".4"
            leftIcon="edit"
            mb={4}
            onClick={[Function]}
            variantColor="yellow"
          >
            <PseudoBox
              _active={
                Object {
                  "bg": "yellow.700",
                }
              }
              _disabled={
                Object {
                  "boxShadow": "none",
                  "cursor": "not-allowed",
                  "opacity": "40%",
                }
              }
              _focus={
                Object {
                  "boxShadow": "outline",
                }
              }
              _hover={
                Object {
                  "bg": "yellow.600",
                }
              }
              alignItems="center"
              appearance="none"
              as="button"
              bg="yellow.500"
              borderRadius="md"
              color="white"
              display="inline-flex"
              fontSize="md"
              fontWeight="semibold"
              height={10}
              justifyContent="center"
              lineHeight="1.2"
              mb={4}
              minWidth={10}
              onClick={[Function]}
              outline="none"
              position="relative"
              px={4}
              transition="all 250ms"
              type="button"
              userSelect="none"
              verticalAlign="middle"
              whiteSpace="nowrap"
              width="full"
            >
              <button
                className="css-emnr3w"
                onClick={[Function]}
                type="button"
              >
                <ButtonIcon
                  icon="edit"
                  ml={-1}
                  mr={2}
                >
                  <Icon
                    aria-hidden="true"
                    color="currentColor"
                    focusable="false"
                    ml={-1}
                    mr={2}
                    name="edit"
                  >
                    <Styled(div)
                      aria-hidden="true"
                      as="svg"
                      color="currentColor"
                      display="inline-block"
                      focusable="false"
                      ml={-1}
                      mr={2}
                      role="presentation"
                      size="1em"
                      verticalAlign="middle"
                      viewBox="0 0 24 24"
                    >
                      <svg
                        aria-hidden="true"
                        className="css-yxiis9"
                        focusable="false"
                        role="presentation"
                        viewBox="0 0 24 24"
                      >
                        <g
                          fill="none"
                          stroke="currentColor"
                          strokeLinecap="full"
                          strokeWidth="2"
                        >
                          <path
                            d="M11 4H4a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7"
                          />
                          <path
                            d="M18.5 2.5a2.121 2.121 0 0 1 3 3L12 15l-4 1 1-4 9.5-9.5z"
                          />
                        </g>
                      </svg>
                    </Styled(div)>
                  </Icon>
                </ButtonIcon>
                Edit route geometry
              </button>
            </PseudoBox>
          </Button>
          <Checkbox
            fontWeight="normal"
            isChecked={true}
            key=".6"
            mb={4}
            onChange={[Function]}
          >
            <Box
              alignItems="center"
              as="label"
              cursor="pointer"
              display="inline-flex"
              fontWeight="normal"
              mb={4}
              verticalAlign="top"
            >
              <label
                className="css-got82x"
              >
                <VisuallyHidden
                  aria-checked={true}
                  as="input"
                  checked={true}
                  onChange={[Function]}
                  type="checkbox"
                >
                  <input
                    aria-checked={true}
                    checked={true}
                    className="css-1gy22ay"
                    onChange={[Function]}
                    type="checkbox"
                  />
                </VisuallyHidden>
                <ControlBox
                  _checked={
                    Object {
                      "bg": "blue.500",
                      "borderColor": "blue.500",
                      "color": undefined,
                    }
                  }
                  _checkedAndDisabled={
                    Object {
                      "bg": "gray.200",
                      "borderColor": "gray.200",
                      "color": "gray.500",
                    }
                  }
                  _disabled={
                    Object {
                      "bg": "gray.100",
                      "borderColor": "gray.100",
                    }
                  }
                  _focus={
                    Object {
                      "boxShadow": "outline",
                    }
                  }
                  _invalid={
                    Object {
                      "borderColor": "red.500",
                    }
                  }
                  alignItems="center"
                  aria-hidden="true"
                  border="2px"
                  borderColor="inherit"
                  color="white"
                  display="inline-flex"
                  flexShrink="0"
                  justifyContent="center"
                  opacity={1}
                  rounded="sm"
                  size={4}
                  transition="background-color 120ms, box-shadow 250ms"
                  userSelect="none"
                >
                  <div
                    aria-hidden="true"
                    className="css-nlndso"
                  >
                    <Icon
                      name="check"
                      size="10px"
                      transition="transform 240ms, opacity 240ms"
                    >
                      <Styled(div)
                        as="svg"
                        color="currentColor"
                        display="inline-block"
                        focusable={false}
                        role="presentation"
                        size="10px"
                        transition="transform 240ms, opacity 240ms"
                        verticalAlign="middle"
                        viewBox="0 0 14 14"
                      >
                        <svg
                          className="css-1vhqk79"
                          focusable={false}
                          role="presentation"
                          viewBox="0 0 14 14"
                        >
                          <g
                            fill="currentColor"
                          >
                            <polygon
                              points="5.5 11.9993304 14 3.49933039 12.5 2 5.5 8.99933039 1.5 4.9968652 0 6.49933039"
                            />
                          </g>
                        </svg>
                      </Styled(div)>
                    </Icon>
                  </div>
                </ControlBox>
                <Box
                  fontSize="md"
                  ml={2}
                  opacity={1}
                  userSelect="none"
                >
                  <div
                    className="css-6qsuox"
                  >
                    Auto-create stops at set spacing
                  </div>
                </Box>
              </label>
            </Box>
          </Checkbox>
          <NumberInput
            key=".7"
            label="Stop spacing"
            mb={4}
            onChange={[Function]}
            test={[Function]}
            units="meters"
            value={3000}
          >
            <FormControl
              isInvalid={false}
              mb={4}
            >
              <Box
                mb={4}
                role="group"
              >
                <div
                  className="css-acwcvw"
                  role="group"
                >
                  <FormLabel
                    htmlFor="00000000-0000-0000-0000-000000000000"
                  >
                    <Box
                      as="label"
                      display="inline-block"
                      fontSize="md"
                      fontWeight="medium"
                      htmlFor="00000000-0000-0000-0000-000000000000"
                      opacity="1"
                      pb="4px"
                      pr="12px"
                      textAlign="left"
                      verticalAlign="middle"
                    >
                      <label
                        className="css-176s5fq"
                        htmlFor="00000000-0000-0000-0000-000000000000"
                      >
                        Stop spacing
                      </label>
                    </Box>
                  </FormLabel>
                  <InputGroup>
                    <Box
                      display="flex"
                      position="relative"
                    >
                      <div
                        className="css-kjafn5"
                      >
                        <Input
                          as="input"
                          errorBorderColor="red.500"
                          focusBorderColor="blue.500"
                          id="00000000-0000-0000-0000-000000000000"
                          isFullWidth={true}
                          isInvalid={false}
                          isValid={true}
                          key=".0"
                          onBlur={[Function]}
                          onChange={[Function]}
                          onFocus={[Function]}
                          pl={null}
                          placeholder=""
                          pr={null}
                          size="md"
                          type="text"
                          value={3000}
                          variant="outline"
                        >
                          <PseudoBox
                            _disabled={
                              Object {
                                "cursor": "not-allowed",
                                "opacity": "0.4",
                              }
                            }
                            _focus={
                              Object {
                                "borderColor": "#3182ce",
                                "boxShadow": "0 0 0 1px #3182ce",
                                "zIndex": 1,
                              }
                            }
                            _hover={
                              Object {
                                "borderColor": "gray.300",
                              }
                            }
                            _invalid={
                              Object {
                                "borderColor": "#e53e3e",
                                "boxShadow": "0 0 0 1px #e53e3e",
                              }
                            }
                            _readOnly={
                              Object {
                                "bg": "transparent",
                                "boxShadow": "none !important",
                                "userSelect": "all",
                              }
                            }
                            alignItems="center"
                            appearance="none"
                            aria-invalid={false}
                            as="input"
                            bg="white"
                            border="1px"
                            borderColor="inherit"
                            display="flex"
                            fontSize="md"
                            height={10}
                            id="00000000-0000-0000-0000-000000000000"
                            isValid={true}
                            onBlur={[Function]}
                            onChange={[Function]}
                            onFocus={[Function]}
                            outline="none"
                            pl={null}
                            placeholder=""
                            position="relative"
                            pr={null}
                            px={4}
                            rounded="md"
                            transition="all 0.2s"
                            type="text"
                            value={3000}
                            width="100%"
                          >
                            <input
                              aria-invalid={false}
                              className="css-1ga4kqw"
                              id="00000000-0000-0000-0000-000000000000"
                              onBlur={[Function]}
                              onChange={[Function]}
                              onFocus={[Function]}
                              placeholder=""
                              type="text"
                              value={3000}
                            />
                          </PseudoBox>
                        </Input>
                        <InputRightElement
                          color="gray.400"
                          key=".1"
                          mr={4}
                          size="md"
                          userSelect="none"
                          width="unset"
                        >
                          <InputElement
                            color="gray.400"
                            mr={4}
                            placement="right"
                            size="md"
                            userSelect="none"
                            width="unset"
                          >
                            <Box
                              alignItems="center"
                              color="gray.400"
                              display="flex"
                              fontSize="md"
                              height={10}
                              justifyContent="center"
                              mr={4}
                              position="absolute"
                              right="0"
                              top="0"
                              userSelect="none"
                              width="unset"
                              zIndex={2}
                            >
                              <div
                                className="css-1he2y93"
                              >
                                meters
                              </div>
                            </Box>
                          </InputElement>
                        </InputRightElement>
                      </div>
                    </Box>
                  </InputGroup>
                </div>
              </Box>
            </FormControl>
          </NumberInput>
          <Checkbox
            fontWeight="normal"
            isChecked={false}
            isDisabled={true}
            key=".8"
            mb={4}
            onChange={[Function]}
          >
            <Box
              alignItems="center"
              as="label"
              cursor="not-allowed"
              display="inline-flex"
              fontWeight="normal"
              mb={4}
              verticalAlign="top"
            >
              <label
                className="css-8syznr"
              >
                <VisuallyHidden
                  aria-checked={false}
                  as="input"
                  checked={false}
                  disabled={true}
                  onChange={[Function]}
                  type="checkbox"
                >
                  <input
                    aria-checked={false}
                    checked={false}
                    className="css-1gy22ay"
                    disabled={true}
                    onChange={[Function]}
                    type="checkbox"
                  />
                </VisuallyHidden>
                <ControlBox
                  _checked={
                    Object {
                      "bg": "blue.500",
                      "borderColor": "blue.500",
                      "color": undefined,
                    }
                  }
                  _checkedAndDisabled={
                    Object {
                      "bg": "gray.200",
                      "borderColor": "gray.200",
                      "color": "gray.500",
                    }
                  }
                  _disabled={
                    Object {
                      "bg": "gray.100",
                      "borderColor": "gray.100",
                    }
                  }
                  _focus={
                    Object {
                      "boxShadow": "outline",
                    }
                  }
                  _invalid={
                    Object {
                      "borderColor": "red.500",
                    }
                  }
                  alignItems="center"
                  aria-hidden="true"
                  border="2px"
                  borderColor="inherit"
                  color="white"
                  display="inline-flex"
                  flexShrink="0"
                  justifyContent="center"
                  opacity={1}
                  rounded="sm"
                  size={4}
                  transition="background-color 120ms, box-shadow 250ms"
                  userSelect="none"
                >
                  <div
                    aria-hidden="true"
                    className="css-nlndso"
                  >
                    <Icon
                      name="check"
                      size="10px"
                      transition="transform 240ms, opacity 240ms"
                    >
                      <Styled(div)
                        as="svg"
                        color="currentColor"
                        display="inline-block"
                        focusable={false}
                        role="presentation"
                        size="10px"
                        transition="transform 240ms, opacity 240ms"
                        verticalAlign="middle"
                        viewBox="0 0 14 14"
                      >
                        <svg
                          className="css-1vhqk79"
                          focusable={false}
                          role="presentation"
                          viewBox="0 0 14 14"
                        >
                          <g
                            fill="currentColor"
                          >
                            <polygon
                              points="5.5 11.9993304 14 3.49933039 12.5 2 5.5 8.99933039 1.5 4.9968652 0 6.49933039"
                            />
                          </g>
                        </svg>
                      </Styled(div)>
                    </Icon>
                  </div>
                </ControlBox>
                <Box
                  fontSize="md"
                  ml={2}
                  opacity={0.4}
                  userSelect="none"
                >
                  <div
                    className="css-wfglws"
                  >
                    Bidirectional
                  </div>
                </Box>
              </label>
            </Box>
          </Checkbox>
          <Alert
            key=".9"
            mb={null}
            status="info"
          >
            <Box
              alignItems="center"
              bg="blue.100"
              display="flex"
              mb={null}
              overflow="hidden"
              pb={3}
              pl={4}
              position="relative"
              pr={4}
              pt={3}
              role="alert"
            >
              <div
                className="css-127nqbe"
                role="alert"
              >
                Modification cannot be bidirectional because timetables in this modification have phasing information.
              </div>
            </Box>
          </Alert>
        </div>
      </Box>
    </Flex>
  </ForwardRef>
</EditAlignment>
`;
